Introduce new 'filetree' lookup plugin

dagwieers opened the pull request:
##### ISSUE TYPE
- Feature Pull Request
##### ANSIBLE VERSION

v2.0
##### SUMMARY

The new "filetree" lookup plugin makes it possible to recurse over a tree of files within the task loop. This makes it possible to e.g. template a complete tree of files to a target system with little effort while retaining permissions and ownership.

The module supports directories, files and symlinks. And also SELinux.

If you provide more than one path, it will implement a `with_first_found` logic, and will not process entries it already processed in previous paths. This enables the user to merge different trees in order of importance, or add e.g. role_vars specific paths to influence different instances of the same role.

The item dictionary consists of:
- src
- root
- path
- mode
- state
- owner
- group
- seuser
- serole
- setype
- selevel
- uid
- gid
- size
- mtime
- ctime
##### EXAMPLES

Here is an example of how we use with_filetree within a role.
The `web/` path is relative to either `roles/<role>/files/` or `files/`.

``` yaml
 - name: Create directories
   file:
     path: /web/{{ item.path }}
     state: directory
     mode: '{{ item.mode }}'
   with_filetree: web/
   when: item.state == 'directory'

 - name: Template files
   template:
     src: '{{ item.src }}'
     dest: /web/{{ item.path }}
     mode: '{{ item.mode }}'
   with_filetree: web/
   when: item.state == 'file'

 - name: Recreate symlinks
   file:
     src: '{{ item.src }}'
     dest: /web/{{ item.path }}
     state: link
     force: yes
     mode: '{{ item.mode }}'
   with_filetree: web/
   when: item.state == 'link'
```
##### SPECIAL USE

The following properties also have its special use:
- root: Makes it possible to filter by original location
- path: Is the relative path to root
- uid, gid: Makes it possible to force-create by exact id, rather than by name
- size, mtime, ctime: Makes it possible to filter out files by size, mtime or ctime

dagwieers commented:
I wonder if this is also eligible for the upcoming 1.9.5 release ?
As we are not allowed to move to 2.0 or 2.1 for a considerable time due to conflicting releases.

dagwieers committed:
Introduce new 'filetree' lookup plugin

The new "filetree" lookup plugin makes it possible to recurse over a tree of files within the task loop. This makes it possible to e.g. template a complete tree of files to a target system with little effort while retaining permissions and ownership.

The module supports directories, files and symlinks.

The item dictionary consists of:
 - src
 - root
 - path
 - mode
 - state
 - owner
 - group
 - seuser
 - serole
 - setype
 - selevel
 - uid
 - gid
 - size
 - mtime
 - ctime

EXAMPLES:
Here is an example of how we use with_filetree within a role:

```yaml
 - name: Create directories
   file:
     path: /web/{{ item.path }}
     state: directory
     mode: '{{ item.mode }}'
     owner: '{{ item.owner }}'
     group: '{{ item.group }}'
     force: yes
   with_filetree: web/
   when: item.state == 'directory'

 - name: Template complete tree
   file:
     src: '{{ item.src }}'
     dest: /web/{{ item.path }}
     state: 'link'
     mode: '{{ item.mode }}'
     owner: '{{ item.owner }}'
     group: '{{ item.group }}'
   with_filetree: web/
   when: item.state == 'link'

 - name: Template complete tree
   template:
     src: '{{ item.src }}'
     dest: /web/{{ item.path }}
     mode: '{{ item.mode }}'
     owner: '{{ item.owner }}'
     group: '{{ item.group }}'
     force: yes
   with_filetree: web/
   when: item.state == 'file'
```

SPECIAL USE:
The following properties also have its special use:

 - root: Makes it possible to filter by original location
 - path: Is the relative path to root
 - uid, gid: Makes it possible to force-create by exact id, rather than by name
 - size, mtime, ctime: Makes it possible to filter out files by size, mtime or ctime

TODO:
 - Add snippets to documentation

dagwieers committed:
Small fixes for Python 3

dagwieers committed:
Return the portion of the file’s mode that can be set by os.chmod()

And remove the exists=True, which is redundant.

dagwieers committed:
Use lstat() instead of stat() since we support symlinks

dagwieers committed:
Avoid a few possible stat() calls

dagwieers committed:
Bring in line with v1.9 and hybrid plugin

dagwieers committed:
Remove glob module since we no longer use it

RussellLuo commented:
:+1:

RussellLuo commented on lib/ansible/plugins/lookup/filetree.py:
```diff
@@ -0,0 +1,122 @@
+        ret['selevel'] = context[3]
+
+    return ret
+
```
Two blank lines will be better. See [PEP 8](https://www.python.org/dev/peps/pep-0008/#blank-lines)

dagwieers replied:
Fixed.

RussellLuo commented on lib/ansible/plugins/lookup/filetree.py:
```diff
@@ -0,0 +1,122 @@
+                for entry in dirs + files:
+                    relpath = os.path.relpath(os.path.join(root, entry), path)
+                    props = file_props(path, relpath)
+                    if props != None:
```
I think `if props is not None` is more conventional :smile:

dagwieers replied:
Fixed.

dagwieers committed:
Included suggestions from @RussellLuo

- Two blank lines will be better. See PEP 8
- I think if props is not None is more conventional :smile:

RussellLuo commented on lib/ansible/plugins/lookup/filetree.py:
```diff
@@ -0,0 +1,123 @@
+from __main__ import display
+warning = display.warning
+
+HAVE_SELINUX=False
```
One space around an assignment operator will be better. Sorry for my OCD of PEP 8 ;-)

dagwieers replied:
This part came directly from the Ansible source code. So if the aim is to make everything PEP8 compliant, I don't mind doing my part. But I don't see a reason for being holier than the pope (dutch expression).

RussellLuo replied:
You are definitely right! For the sake of consistency, PEP8-incompliance is never a problem.

I won't give this advice if I were more familiar with the Ansible source code. Sorry for my self-approbation ;-(

RussellLuo commented on lib/ansible/plugins/lookup/filetree.py:
```diff
@@ -0,0 +1,123 @@
+HAVE_SELINUX=False
+try:
+    import selinux
+    HAVE_SELINUX=True
```
one space around...

RussellLuo commented on lib/ansible/plugins/lookup/filetree.py:
```diff
@@ -0,0 +1,123 @@
+    HAVE_SELINUX=True
+except ImportError:
+    pass
+
```
Two blank lines between two top-level functions, sorry for not explicitly pointing this out before.

RussellLuo commented on lib/ansible/plugins/lookup/filetree.py:
```diff
@@ -0,0 +1,123 @@
+    if isinstance(path, unicode):
+        path = path.encode("utf-8")
+    return path
+
```
two blank lines...

RussellLuo commented on lib/ansible/plugins/lookup/filetree.py:
```diff
@@ -0,0 +1,123 @@
+        path = path.encode("utf-8")
+    return path
+
+# If selinux fails to find a default, return an array of None
```
And move this comment line as the function docstring of `selinux_context` will looks better ;-)

dagwieers replied:
This comes directly from the Ansible source code. Again, if this is general advice, let's fix it in Ansible first then get back to me with your PEP8 OCD.

RussellLuo replied:
I can't agree more! Please ignore these immature suggestions :smile:

dagwieers committed:
Support failed pwd/grp lookups

dagwieers committed:
Implement first-found functionality in the path-order

dagwieers commented:
shipit

dagwieers commented:
Or rather :shipit:

dagwieers commented:
@jimi-c @bcoca How about merging this one for v2.1.0 ?

jimi-c commented:
@dagwieers sorry this didn't make the cut, adjusted the milestone so we can grab it for 2.2.

DanyC97 commented:
@jimi-c 2.2 is out however i don't see it documented. I've checked the stable-2.2 branch and i can see the code so i assume the doc is a bit behind ?

austinbutler commented:
This seems to work well for templates, except the files retain the .j2 extension. Is that intended? If so, would it be possible to have an option to strip that?

dagwieers commented:
You can do that using jinja2. There is no reason to add functionality in individual modules if you can do the same using jinja2 filters. 

- http://jinja.pocoo.org/docs/2.9/templates/#builtin-filters
- http://docs.ansible.com/ansible/playbooks_filters.html

So something like: `{{ item.src | splitext | first }}` would do.
Or `{{ item.src | replace('.j2', '') }}`.
Or even: `{{ item.src | regexreplace('.j2$', '') }}`

I haven't tested these, it's a bit theoretical. Let me know if this works, would be nice to add this to the documentation at some point. Thanks !

austinbutler commented:
Sorry, edited my comment on you but `regex_replace('.j2','')` worked for me. Having it in the documentation would be great, thanks!

dagwieers commented:
@austinbutler I would still add the `$` in that regex, just in case.

austinbutler commented:
You're right, that should be safer. Added and tested again, all good.

splashx commented:
Is this on 2.3.0.0? Using devel, can't run `lookup('filetree', 'dir/')` nor `with_filetree`

dagwieers commented:
Documentation added here: #26157

JPvRiel commented:
> You can do that using jinja2. There is no reason to add functionality in individual modules/plugins if you can do the same using jinja2 filters.

Doesn't the `template` module strip `.j2` when used with a single template file? AFAIK, the current `template` module behaviour with a normal `.j2` extension source file is to strip the `.j2` extension for you. 

So, I disagree with the argument that jinja2 filters should be used because context matters. If it was the `copy` or `fetch` module, that extra regex filter makes sense. But with `template`, why must a user expect and code around different behaviour when `with_filetree` versus normal files? That seems inconsistent to me and I bet many other users of `with_filetree` would find in non-intuitive if the `.j2` extension doesn't get stripped.

dagwieers commented:
@JPvRiel The **filetree** plugin returns a dictionary with information for each file, that is it. How users use it, is up to the user. I don't see what the **filetree** plugin should do different here.

PS AFAICT the **template** module does not strip anything from the provided filenames.

JPvRiel commented:
> PS AFAICT the template module does not strip anything from the provided filenames.

Sorry, yip, my bad - I double checked now and indeed, it doesn't strip the `.j2` extension like I thought. Your original point about using jinja2 filters is valid.
